pkg/network/dhcp/server/server.go

```diff
@@ -205,6 +205,11 @@ func (h *DHCPHandler) ServeDHCP(p dhcp.Packet, msgType dhcp.MessageType, _ dhcp.
 
 	case dhcp.Request:
 		log.Log.V(4).Info("The request has message type REQUEST")
+		if requestIP, ok := opts[dhcp.OptionRequestedIPAddress]; ok && !net.IP(requestIP).Equal(h.clientIP) {
+			log.Log.Warningf("Client cannot request ip %v, not the assigned ip %v", net.IP(requestIP), h.clientIP)
+			return dhcp.ReplyPacket(p, dhcp.NAK, h.serverIP, nil, h.leaseDuration,
+				h.options.SelectOrderOrAll(nil))
+		}
```

Comment:
Interesting. So this would be a scenario where a guest VM got assigned an IP from one run of the VM. Then the VM was shutdown and started again in a new pod, but somehow retained knowledge of the previous IP?

What OS does this? If this is something that you can reproduce consistently, it's possible we can add this scenario to our test suite

Reply:
Thanks for the proposed fix.
When you performed the reboot, it was using `virctl` or directly from the VM's console.

+1 to David's question. What OS the VM has? What DHCP client it has?

Also, iiuc, if we get to the point the DHCP server is sending a NACK (the new added code) it means the client sent a DHCP request. Previously we just sent a reply with the correct IP. Can you please explain why the NACK is needed and a reply with the correct IP is not enough?

Reply:
1. I reboot VM through patch VM spec to `running: false`, then patch `running: true`
2. OS is `windows server 2019`
3. I used default dhcp client which is built-in the `windows server 2019`
4.  dhcp server sends NACK since we found request IP doesn't equal to client IP, means we need VM to re-request an new IP

Reply:
According to [this](https://www.juniper.net/documentation/en_US/junos/topics/concept/extended-dhcp-local-server-option-50-acx-series.html#:~:text=Subscriber%20management%20enables%20you%20to,using%20prior%20to%20being%20disconnected.), the server may ignore the requested IP and propose another IP. That's what the old code done. I guess that the windows DHCP client ignores the new proposal and NACK is indeed what solves the issue (NACK is the proposed solution in [this](https://docs.microsoft.com/en-us/windows-server/troubleshoot/dynamic-host-configuration-protocol-basics) microsoft doc).

Before approving this PR we need to verify (better by adding e2e tests:)) that this option work for linux guest as well and doesn't break their networking.